include/exadg/incompressible_navier_stokes/spatial_discretization/operators/convective_operator.h

```diff
@@ -745,51 +828,118 @@ class ConvectiveKernel
 
-    scalar delta_average_normal_velocity = delta_average_velocity * normalM;
-
     vector jump_value = delta_uM - delta_uP;
 
-    return (average_normal_velocity * delta_average_velocity +
-            delta_average_normal_velocity * average_velocity +
-            data.upwind_factor * 0.5 * std::abs(average_normal_velocity) * jump_value);
+    vector flux;
+    if(data.linearization_type == LinearizationType::Newton)
+    {
+      scalar delta_average_normal_velocity = delta_average_velocity * normalM;
+
+      flux = average_normal_velocity * delta_average_velocity +
+             delta_average_normal_velocity * average_velocity +
+             (data.upwind_factor * 0.5 * std::abs(average_normal_velocity)) * jump_value;
+    }
+    else if(data.linearization_type == LinearizationType::Picard)
+    {
+      flux = average_normal_velocity * delta_average_velocity +
+             (data.upwind_factor * 0.5 * std::abs(average_normal_velocity)) * jump_value;
+    }
+    else
+    {
+      AssertThrow(false, dealii::ExcMessage("Linearization type not implemented."));
+    }
+
+    return flux;
   }
 
   /*
    * Velocity:
    *
-   *  Linearized convective operator (= homogeneous operator):
-   *  Dirichlet boundary: delta_u⁺ = - delta_u⁻ or 0
+   *  Linearized convective operator
+   *  homogeneous operator for Newton linearization (g = 0)
+   *  inhomogeneous operator for Picard linearization
+   *
+   *  Dirichlet boundary: delta_u⁺ = - delta_u⁻ + 2 g or g
    *  Neumann boundary:   delta_u⁺ = + delta_u⁻
    *  symmetry boundary:  delta_u⁺ = delta_u⁻ - 2 (delta_u⁻*n)n
+   *
    */
   inline DEAL_II_ALWAYS_INLINE //
     dealii::Tensor<1, dim, dealii::VectorizedArray<Number>>
     calculate_exterior_value_linearized(
       dealii::Tensor<1, dim, dealii::VectorizedArray<Number>> & delta_uM,
       unsigned int const                                        q,
       FaceIntegrator<dim, dim, Number> &                        integrator,
-      BoundaryTypeU const &                                     boundary_type) const
+      OperatorType const &                                      operator_type,
+      BoundaryTypeU const &                                     boundary_type,
+      dealii::types::boundary_id const                          boundary_id,
+      std::shared_ptr<BoundaryDescriptorU<dim> const>           boundary_descriptor,
+      double const &                                            time) const
   {
     // element e⁺
     dealii::Tensor<1, dim, dealii::VectorizedArray<Number>> delta_uP;
 
     if(boundary_type == BoundaryTypeU::Dirichlet || boundary_type == BoundaryTypeU::DirichletCached)
     {
-      if(data.type_dirichlet_bc == TypeDirichletBCs::Mirror)
+      if(operator_type == OperatorType::full || operator_type == OperatorType::inhomogeneous)
       {
-        delta_uP = -delta_uM;
+        dealii::Tensor<1, dim, dealii::VectorizedArray<Number>> g;
+
+        if(data.linearization_type == LinearizationType::Newton)
```

Comment:
I did not write 2 different functions just because of Picard/Newton, since they only differ in how they treat the boundary data g. Here, Newton uses g=0 (homogeneous operator), whereas Picard uses the actual g.